File: 2021/d4/rust/src/main.rs
use std::path::Path;
use std::fs::File;
use std::io::Read;
use itertools::Itertools;

fn calculate_score(board: &Vec<Vec<(usize, bool)>>) -> usize {
    let mut score = 0;

    for row in board {
        for (num, marked) in row {
            if !marked {
                score += num;
            }
        }
    }

    score
}

fn winning_board(board: &Vec<Vec<(usize, bool)>>) -> bool {
    let mut cols = vec![true; 5];

    for row in board {
        if row.iter().all(|(_, marked)| *marked) {
            return true;
        }

        for (ref mut col, (_, marked)) in cols.iter_mut().zip(row.iter()) {
            **col &= marked
        }
    }

    if cols.iter().any(|c| *c) {
        true
    } else {
        false
    }
}

fn bingo(calls: &Vec<usize>, boards: &Vec<Vec<Vec<(usize, bool)>>>) -> usize {
    let mut boards = boards.clone();

    for call in calls {
        // mark on all the boards
        for board in boards.iter_mut() {
            for row in board {
                for (num, ref mut marked) in row {
                    if num == call {
                        *marked = true
                    }
                }
            }
        }

        // check if any boards have won
        for board in boards.iter() {
            if winning_board(&board) {
                return calculate_score(&board) * call;
            }
        }
    }

    panic!("no winning boards");
}

fn lose_bingo(calls: &Vec<usize>, boards: &Vec<Vec<Vec<(usize, bool)>>>) -> usize {
    let mut boards = boards.clone();

    for call in calls {
        // mark on all the boards
        for board in boards.iter_mut() {
            for row in board {
                for (num, ref mut marked) in row {
                    if num == call {
                        *marked = true
                    }
                }
            }
        }

        if boards.len() == 1 {
            if winning_board(&boards[0]) {
                return calculate_score(&boards[0]) * call;
            }
        } else {
            // check if any boards have won
            boards = boards.into_iter().filter(|board| {
                !winning_board(&board)
            }).collect();
        }
    }

    panic!("no winning boards");
}

fn main() {
    let mut args = std::env::args();

    let input_filename = args.nth(1).unwrap_or_else(|| {
        eprintln!("No input file specified");
        eprintln!("Usage: ./run <input filename>");
        std::process::exit(1);
    });

    let mut input_file = File::open(Path::new(&input_filename)).expect("failed to open input file");
    let mut input = String::new();
    input_file.read_to_string(&mut input).expect("failed to read from file");

    let mut lines = input.lines();

    let calls = lines.next().expect("no line of calls")
        .split(',').map(|c| c.parse().expect("failed to parse a number"))
        .collect();
    let _ = lines.next().expect("no blank line");

    let boards: &Vec<Vec<Vec<(usize, bool)>>> = &lines.chunks(6).into_iter().map(|board| {
        board.take(5)
            .map(|l| l.split_ascii_whitespace().map(|n| (n.parse().expect("failed to parse board number"), false)).collect())
            .collect()
    }).collect();

    println!("part 1: {}", bingo(&calls, boards));
    println!("part 2: {}", lose_bingo(&calls, boards));
}

File: 2021/d4/python/main.py
import sys
import re


class Board:
    def __init__(self, data):
        self.data = data
        self.selected = list(map(lambda x: [False] * len(x), data))

    def select(self, guess):
        for y in range(len(self.data)):
            for x in range(len(self.data[y])):
                if self.data[y][x] == guess:
                    self.selected[y][x] = True
                    return

    def winner(self):
        for row in self.selected:
            if all(row):
                return True

        for col in range(len(self.selected[0])):
            if all(map(lambda x: x[col], self.selected)):
                return True

        return False

    def score(self, last_guess):
        unselected_sum = 0

        for y in range(len(self.data)):
            for x in range(len(self.data[y])):
                if not self.selected[y][x]:
                    unselected_sum += self.data[y][x]

        return unselected_sum * last_guess


def winning_score(guesses, boards):
    for guess in guesses:
        for board in boards:
            board.select(guess)

        for board in boards:
            if board.winner():
                return board.score(guess)

def losing_score(guesses, boards):
    for guess in guesses:
        for board in boards:
            board.select(guess)

        if len(boards) > 1:
            boards = [board for board in boards if not board.winner()]
        else:
            if boards[0].winner():
                return boards[0].score(guess)


if __name__ == '__main__':
    if len(sys.argv) < 2:
        raise Exception("filename not speciifed, specify filename as first argument")

    filename = sys.argv[1]
    lines = None

    with open(filename) as f:
        lines = [line.strip() for line in f]

    if not lines:
        raise Exception("no lines found")

    guesses = [int(guess) for guess in lines.pop(0).split(",")]
    _ = lines.pop(0) # get rid of the blank line before the boards

    boards = []
    board = []

    for line in lines:
        if line == "":
            boards.append(Board(board))
            board = []
        else:
            board.append([int(n.strip()) for n in re.split('\s+', line)])

    boards.append(Board(board))

    print("part 1: " + str(winning_score(guesses, boards)))
    print("part 2: " + str(losing_score(guesses, boards)))
